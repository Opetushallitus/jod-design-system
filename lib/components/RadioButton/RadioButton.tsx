import { Radio } from '@headlessui/react';

export interface RadioButtonProps {
  /** Text for the component */
  label: string;
  /** Value for the component */
  value: string;
  /** CSS classes for custom styles */
  className?: string;
}

export const RadioButton = ({ label, value, className }: RadioButtonProps) => {
  return (
    <Radio value={value} className={`${className ? className : ''} flex h-7`.trim()}>
      {({ checked }) => (
        <div className="ds-flex-start ds-flex ds-space-x-4">
          {checked ? <CheckedIcon /> : <UncheckedIcon />}
          <span className="ds-font-poppins ds-flex ds-items-center ds-text-button-md ds-text-black hover:ds-text-accent hover:ds-underline ds-hyphens-auto">
            {label}
          </span>
        </div>
      )}
    </Radio>
  );
};

const CheckedIcon = () => {
  return (
    <svg
      xmlns="http://www.w3.org/2000/svg"
      width="16"
      height="16"
      viewBox="0 0 16 16"
      fill="none"
      className="ds-self-center"
    >
      <circle cx="8" cy="8" r="8" className="ds-fill-accent" />
    </svg>
  );
};

const UncheckedIcon = () => {
  return (
    <svg
      xmlns="http://www.w3.org/2000/svg"
      width="16"
      height="16"
      viewBox="0 0 16 16"
      fill="none"
      className="ds-self-center"
    >
      <circle cx="8" cy="8" r="8" className="ds-fill-white" />
    </svg>
  );
};
