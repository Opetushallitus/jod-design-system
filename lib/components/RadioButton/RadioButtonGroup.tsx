import React from 'react';
import { Label, RadioGroup } from '@headlessui/react';

export interface RadioButtonGroupProps {
  /** Text for the component */
  label: string;
  /** Value that is currently selected */
  value: string;
  /** Callback function when the value changes */
  onChange: (newValue: string) => void;
  /** Use RadioButton components as children */
  children: React.ReactNode;
  /** CSS classes for custom styles */
  className?: string;
}

export const RadioButtonGroup = ({ label, value, onChange, children, className }: RadioButtonGroupProps) => {
  return (
    <RadioGroup
      value={value}
      onChange={onChange}
      className={`${className ? className : ''} flex flex-col space-y-2`.trim()}
    >
      <Label className="mb-5 text-body-xs text-secondary-gray">{label}</Label>
      {children}
    </RadioGroup>
  );
};
