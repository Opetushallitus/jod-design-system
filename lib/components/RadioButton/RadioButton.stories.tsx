import type { StoryObj } from '@storybook/react-vite';
import React from 'react';
import { fn } from 'storybook/test';
import type { TitledMeta } from '../../utils';

import { RadioButton } from './RadioButton';
import { RadioButtonGroup, RadioButtonGroupProps } from './RadioButtonGroup';

const meta = {
  title: 'Forms/RadioButton',
  component: RadioButtonGroup,
  tags: ['autodocs'],
} satisfies TitledMeta<typeof RadioButtonGroup>;

export default meta;

type Story = StoryObj<typeof meta>;

const parameters = {
  design: {
    type: 'figma',
    url: 'https://www.figma.com/file/6M2LrpSCcB0thlFDaQAI2J/cx_jod_client?node-id=746-43288',
  },
};

const DefaultStoryRender = (props: RadioButtonGroupProps) => {
  const [value, setValue] = React.useState<string>('a');
  return <RadioButtonGroup {...props} value={value} onChange={setValue} />;
};

export const Default: Story = {
  parameters: {
    ...parameters,
  },
  argTypes: {
    children: {
      control: false,
    },
  },
  args: {
    label: 'Valitse, miten haluat ryhmitell채 osaamisesi.',
    value: 'a',
    onChange: fn(),
    children: (
      <>
        <RadioButton label="L채hteiden mukaan" value="a" />
        <RadioButton label="Teemoittain" value="b" />
        <RadioButton label="Aakkosellisesti" value="c" disabled={true} />
      </>
    ),
  },
  render: DefaultStoryRender,
};

export const WithOneDisabled: Story = {
  parameters: {
    ...parameters,
    backgrounds: {
      default: 'jod-bg-white',
      values: [{ name: 'jod-bg-white', value: '#fff' }],
    },
  },
  argTypes: {
    children: {
      control: false,
    },
  },
  args: {
    label: 'Valitse, miten haluat ryhmitell채 osaamisesi.',
    value: 'a',
    onChange: fn(),
    children: (
      <>
        <RadioButton label="L채hteiden mukaan" value="a" />
        <RadioButton label="Teemoittain" value="b" />
        <RadioButton label="Aakkosellisesti" value="c" disabled />
      </>
    ),
  },
  render: DefaultStoryRender,
};
