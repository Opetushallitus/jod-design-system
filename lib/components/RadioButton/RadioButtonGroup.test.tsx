import { render, screen } from '@testing-library/react';
import { describe, expect, it, test, vi } from 'vitest';
import userEvent from '@testing-library/user-event';
import { RadioButtonGroup } from './RadioButtonGroup';
import { RadioButton } from './RadioButton';
import '@testing-library/jest-dom/vitest';

describe('Snapshot testing', () => {
  test('Default', () => {
    const { container } = render(
      <RadioButtonGroup label="A" value="a" onChange={vi.fn()}>
        <RadioButton value="option1" label="Option 1" />
      </RadioButtonGroup>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('RadioButtonGroup', () => {
  it('renders the label correctly', () => {
    render(
      <RadioButtonGroup label="Test Label" value="" onChange={vi.fn()}>
        <RadioButton value="option1" label="Option 1" />
      </RadioButtonGroup>,
    );

    expect(screen.getByText('Test Label')).toBeInTheDocument();
  });

  it('renders the children correctly', () => {
    render(
      <RadioButtonGroup label="Test Label" value="" onChange={vi.fn()}>
        <RadioButton value="option1" label="Option 1" />
      </RadioButtonGroup>,
    );

    expect(screen.getByText('Option 1')).toBeInTheDocument();
  });

  it('calls the onChange callback when a RadioButton is selected', async () => {
    const user = userEvent.setup();
    const mockOnChange = vi.fn();

    render(
      <RadioButtonGroup label="Test Label" value="" onChange={mockOnChange}>
        <RadioButton value="option1" label="Option 1" />
      </RadioButtonGroup>,
    );

    const radioButton = screen.getByText('Option 1');
    await user.click(radioButton);
    expect(mockOnChange).toHaveBeenCalledTimes(1);
  });
});
