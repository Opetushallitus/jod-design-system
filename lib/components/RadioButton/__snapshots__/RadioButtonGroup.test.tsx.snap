// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Snapshot testing > Default 1`] = `
<div
  aria-labelledby="headlessui-label-:r1:"
  class="ds-flex ds-flex-col ds-space-y-2"
  id="headlessui-radiogroup-:r0:"
  role="radiogroup"
>
  <div
    class="ds-mb-5 ds-text-body-xs ds-text-secondary-gray"
    data-headlessui-state=""
    id="headlessui-label-:r1:"
  >
    A
  </div>
  <span
    aria-checked="false"
    class="flex h-7"
    data-headlessui-state=""
    id="headlessui-radio-:r2:"
    role="radio"
    tabindex="0"
  >
    <div
      class="ds-flex-start ds-flex ds-space-x-4"
    >
      <svg
        class="ds-self-center"
        fill="none"
        height="16"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          class="ds-fill-white"
          cx="8"
          cy="8"
          r="8"
        />
      </svg>
      <span
        class="ds-font-poppins ds-flex ds-items-center ds-text-button-md ds-text-black hover:ds-text-accent hover:ds-underline ds-hyphens-auto"
      >
        Option 1
      </span>
    </div>
  </span>
</div>
`;
